compatibility.bs

```diff
@@ -812,6 +812,84 @@ The following are the event handlers and their corresponding event handler event
 WebKit also has this on <a href="https://github.com/WebKit/webkit/blob/e455672f9e6861ced85d8be01cb7bc03a30a0555/LayoutTests/fast/dom/event-handler-attributes.html#L335">HTMLFrameSetElement</a>. It's unclear if this is needed for compatibility.
 </div>
 
+<h3 id="animation-event-handlers">Animation Events</h3>
+
+<h4 id="animation-prefixed-event-handlers-on-elements-document-objects-and-window-objects">Prefixed event handlers on elements, Document objects, and Window objects</h4>
+
+The following are the <a>event handlers</a> (and their corresponding <a>event
+handler event types</a>) that must be supported by all <a>HTML elements</a>, as
+both <a>event handler content attributes</a> and <a>event handler IDL
+attributes</a>; and that must be supported by all {{Document}} and {{Window}}
+objects, as event handler IDL attributes:
+
+<table class="event-handlers" dfn-type=attribute dfn-for="Document, Window">
+  <tr>
+    <th><a>Event handler</a></th>
+    <th><a>Event handler event type</a></th>
+  </tr>
+  <tr>
+    <td><dfn>onwebkitanimationstart</dfn></td>
+    <td>{{animationstart}}</td>
```

Comment:
I feel like these probably need to use the event types defined in https://dom.spec.whatwg.org/#concept-event-listener-invoke 8.2? Or maybe it is not observable which one you choose, hmm... @annevk?

Reply:
If the listener is for `animationstart` an event with type `animationstart` when dispatched will also be `animationstart` for that listener and not corrected to `webkitAnimationStart`. I don't really know what implementations do.

Tests for that would be good, including whether they might alias these IDL attributes.

Reply:
I'm not sure exactly what you're looking for here. Let me run through some things I've learned from testing. Note that implementations (Chrome, Safari, Firefox) have been entirely consistent so far across my testing, which is nice at least :).

i. Triggering a real `animationstart` event will cause prefixed event listeners to receive type=webkitAnimationStart and unprefixed event listeners to receive type=animationstart. You can see this by setting the listeners on different targets in the bubble-chain, to avoid the squashing behavior from https://dom.spec.whatwg.org/#concept-event-listener-invoke step 8.

ii. Sending a fake `animationstart` event (i.e. `window.dispatchEvent(new AnimationEvent('animationstart'));`) does not trigger a prefixed event listener (whether an event handler or directly added listener).

iii. Sending a fake `webkitAnimationStart` event (i.e. `window.dispatchEvent(new AnimationEvent('webkitAnimationStart'));` does trigger a prefixed event listener, and will not trigger an unprefixed one.

iv. Setting `window.onanimationstart = () => { console.log('foo'); }` does not affect the value of `window.onwebkitanimationstart` (it's still null) and vice-versa. They do not seem to be aliases from that perspective.

Do those cover the cases you were thinking of and indicate which way we should go?

Reply:
Thank you! I'd love to see those turned into web-platform-tests. They indeed cover the various scenarios. (Part of what I was getting at is that the text proposed here seems incorrect in that `onwebkitanimationstart` should probably be for `webkitanimationstart` and not `animationstart`.)

ii is unexpected and not supported by the current standard. Synthetic ("fake") events don't go down a different code path. I suppose we could branch on isTrusted in the standard though to limit the weirdness if that's effectively what implementations are doing. That would require a change to whatwg/dom.

Reply:
I think (ii) is supported by the current standard, it's just that the phrasing isn't quite precise so @annevk might have gotten confused. In particular https://dom.spec.whatwg.org/#concept-event-listener-invoke step 8 checks `isTrusted` so `window.dispatchEvent(new AnimationEvent('animationstart'));` will not hit step 8.

Reply:
Well, it only took about 2 months, but I finally got some time (on, err, Christmas Eve) to work on those Web Platform Tests. I think I've covered the cases @annevk was curious about. The updated versions are at https://github.com/web-platform-tests/wpt/pull/19099/files, and the results are https://wpt.fyi/results/compat?run_id=373500001&run_id=381460002&run_id=377730003

Chrome fails a bunch of them because it doesn't define the `on{animation,webkit}XYZ` handlers on `Element` (just `window`). Once that is fixed (moved to GlobalEventHandlers), Chrome should start passing more (most?).

Safari is timing out, I've not yet investigated why this is at all.

There was one bit of surprising behavior that I found; I didn't expect that a prefixed event handler would be deduplicated against an unprefixed event *listener* (and vice-versa). So if you have `onwebkitanimationstart` and `addEventListener('animationstart', ...)`, the `onwebkitanimationstart` handler won't trigger.

What I haven't paged back in yet is all the details about what spec changes are actually needed 😆 . Will try and do that next.

Reply:
I've made some improvements to the tests, latest runs are https://wpt.fyi/results/compat?run_id=398290001&run_id=381650003&run_id=387230001. Seems to reveal a few small issues:

- Firefox is not treating the type passed to prefixed event listeners as fully case sensitive; it accepts (e.g.) `webkittransitionend` (lowercase) when it should only accept `webkitTransitionEnd`.
- For webkitTransitionEnd only, Firefox seems not to be triggering both the prefixed handler + listener when `div.dispatchEvent(new AnimationEvent('webkitTransitionEnd'))` is called. Weirdly, it works for all the webkitAnimationXX types.
- Chromium has the handlers in the wrong place (https://crbug.com/999895). I've confirmed that fixing that makes Chromium pass all the tests (https://chromium-review.googlesource.com/c/chromium/src/+/1986885)
- Safari passes all the tests, though I had to add in a small hack to avoid https://bugs.webkit.org/show_bug.cgi?id=205732

Reply:
Thanks Stephen!

Reply:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1607918 for the case sensitivity. I'm having trouble reproducing the other one, I think I'll wait for the tests to land and point to them.